### ISSUE TITLE
"Community-approved collections" instead of file inclusion in the Ansible package.

### ISSUE THREAD

**Issue by Andersson007** [ansible-community/community-topics#198]
### Summary

When reading the meeting [minutes](https://meetbot.fedoraproject.org/ansible-community/2023-02-15/ansible_community_meeting.2023-02-15-19.00.log.html), the community came back to the question summarized by @samccann:

> 19:41:08 <samccann> so the meta discussion is - what is the future of Ansible community package? Do we let it keep growing and growing? And do we expand it out to collections of roles/playbooks?

I'd like to propose discussing the following. How about:

- We'll stop [including](https://github.com/ansible-collections/ansible-inclusion/blob/main/README.md) new collections in the Ansible package. Instead:
- We'll introduce `Community-approved collections`, so the inclusion process we currently use will be not about collection inclusion but about the **Community Approval**.
- The `Community-approved` status could imply:
  - Appearing the collection on docs.ansible.com in a special list of Community-approved collections
  - Their docs will be on docs.ansible.com (as is now)
  - There can be a special label / section for such collection on Galaxy
- Additionally (don't focus on this detail), ~~there can be a mechanism developed: if a user uses a FQCN of community-approved collection in a playbook, if the collection is not installed yet, core will try to install it from Galaxy automatically.~~ EDITED: let's not bring up such things for now as it's insufficient
- ~~The Ansible package will be scheduled for deprecation.~~ EDITED: we'll stop inclusion and maybe in far future we'll deprecate the Ansible package with a long transition period. But let's not focus on deprecation right now.
- The inclusion process will still work but will be renamed `Community approval` or something. So we'll keep the motivation we currently have.

Thoughts?

Related discussions for reference:
- [Role or playbook-focused collections: include in the Ansible or not?](https://github.com/ansible-community/community-topics/issues/197)
- [Future of Ansible package](https://github.com/ansible-community/community-topics/issues/82)
- [Community meeting 2023-02-15 meeting minutes](https://meetbot.fedoraproject.org/ansible-community/2023-02-15/ansible_community_meeting.2023-02-15-19.00.log.html)

EDITED: changed "Community certification" to "Community approval"

**Comment by mariolenz** [comment]
I see the problem there. But I like the community package as it is. We (like a lot of other environments) don't allow downloading and installing things directly from the internet. We need something in between. (Don't even THINK about discussing security with me, it's a compliance thing!)

So there are tools like [Nexus](https://www.sonatype.com/products/nexus-repository) or [Artifactoy](https://jfrog.com/artifactory/) or similar to mirror / proxy repositories. PyPI is widely supported by those tools which makes it easy to install the community package. Unfortunately, Galaxy isn't.

For example, there's [Add Ansible Galaxy Proxy support](https://issues.sonatype.org/browse/NEXUS-24270) but AFAIK Sonatype didn't implement this yet.

The more we see support for Galaxy in those tools, the easier it is to deprecate and drop the Ansible package.

**Comment by felixfontein** [comment]
Regarding Artifactory, there's @briantist's [galactory](https://github.com/briantist/galactory) project.

**Comment by mariolenz** [comment]
> Regarding Artifactory, there's @briantist's [galactory](https://github.com/briantist/galactory) project.

This is really great, but do people get support from JFrog for this? That's what I'm aiming at: More **official** support for Galaxy would help.

**Comment by briantist** [comment]
> > Regarding Artifactory, there's [galactory](https://github.com/briantist/galactory) project.
> 
> This is really great, but do people get support from JFrog for this? That's what I'm aiming at: More **official** support for Galaxy would help.

If JFrog added support I wouldn't have had to create this. 

Galaxy support was first requested in like 2016 before collections even existed but the request was closed. 

Collection support was requested in 2019 abs the request is still open with no sign they will implement it.

I can link the specific tickets later. 

---

Re: the original topic though, I can't read the minutes right now, is there some reason I'm missing why we should be trying to deprecate the package? 

The proposal for community certified collections as a replacement seems to presuppose the removal of the package but I don't understand why. 

If there's context in the meeting minutes could we please add it to this discussion?

**Comment by mariolenz** [comment]
> If JFrog added support I wouldn't have had to create this.

This is just my point. In some environments, it's relatively easy to install the community package via some existing PyPI mirror. But the tools allowing this usually don't allow the same thing for Galaxy, at least not out of the box. This makes it difficult for some (a lot?) of people to install collections.

I'm not saying this is a reason to _not_ deprecate the Ansible package, but I think this is an important thing we should consider.

> Collection support was requested in 2019 and the request is still open with no sign they will implement it.
> 
> I can link the specific tickets later.

[This one](https://jfrog.atlassian.net/browse/RTFACT-20684)?

**Comment by briantist** [comment]
> [This one](https://jfrog.atlassian.net/browse/RTFACT-20684)?

yes, that's the newer collections-based one, and it links to the older one that was only for roles

---

Ok so after reading the minutes, and the linked discussions, I am re-adding a question I edited out because I couldn't follow up (traveling): why are we talking about deprecating the ansible package at all? As pointed out in the minutes, we have had this discussion before, and it is currently closed:
- https://github.com/ansible-community/community-topics/issues/82

It's been less than a year since it was opened, and less than 6 months since it was closed. I don't think much has changed since then but if we want to revisit that discussion, we should either re-open that topic, or open a new one. 

Until / unless that happens, I find it really difficult to weigh in on this certified-collection proposal.

**Comment by felixfontein** [comment]
I am against deprecating the [Ansible community package](https://pypi.org/project/ansible/) anytime soon. This also was never really discussed during last week's meeting.

Regarding the idea of community-certified collections: this sounds like a interesting idea. And one I think can work fine in parallel to the Ansible package. It seems that quite a few collections ask for inclusion in Ansible also because they want to get more visibility by being included on docs.ansible.com/ansible/latest/. But giving visibility shouldn't be related to being included in Ansible. (Also it will reduce package bloat.)

One big question though is a) how this should work, and b) who does all the certifications. Regarding b), collection reviews take a lot of time. We are already struggling to get enough reviews for collections that want to be included in Ansible, so opening another can of worms^Hreview request sources sounds problematic. Regarding a), it will be important to clarify which of the documented collections will be included in Ansible, and which are just listed "because they are certified". We cannot simply just list them on https://docs.ansible.com/ansible/latest/collections/ and include their content in the plugin/role/module indexes.

TLDR: I like the idea, but it needs more work. But IMO it's a good idea to start discussing it now and keeping it in the backs of our minds, so we can remove these obstacles eventually.

**Comment by Andersson007** [comment]
>I am re-adding a question I edited out because I couldn't follow up (traveling): why are we talking about deprecating the ansible package at all? As pointed out in the minutes, we have had this discussion before, and it is currently closed

>It's been less than a year since it was opened, and less than 6 months since it was closed. I don't think much has changed since then but if we want to revisit that discussion, we should either re-open that topic, or open a new one.
Until / unless that happens, I find it really difficult to weigh in on this certified-collection proposal.

@briantist I think this is the new one:) For the context: this is my personal idea if it's important.
The issues now imo are (or can be in future):

- possible package bload: the package contains the whole Galaxy if everyone decides submit their roles for inclusion
- many collections need just to be visible, not necessarily be physically included
- many collections need just status "we are included, we are quality and cool"
- so visibility via presence on docs.ansible.com in the list of "community-approved/ertified" collections and having their docs there, maybe also labels on galaxy, feels like a solution.

So, imo, by certification instead of inclusion we and maintainers can achieve the following:
- recognition: via getting the status
- motivation: creators/maintainers will still be motivated to create quality stuff; at least i think so:)
- the package will not grow 

There was also a hot discussion about "to include or not to include new collections in the package".
The suggested certification model will eliminate the need to include new stuff.
We will review submitted collections for certification, not for inclusion. The process will be the same as for inclusion, we'll just change `inclusion` -> `community certification` in the repo's name and everywhere. 

>I am against deprecating the Ansible community package anytime soon. This also was never really discussed during last week's meeting.

@felixfontein I'm not talking about deprecating **soon**. The package can continue to exist for compatibility reasons. Though at some point it can be deprecated (with the long transition period) and become like an alias for `ansible-core` (maybe?) but it's another topic.
The key thing is, for now, that we will stop including new stuff in the package.

>But giving visibility shouldn't be related to being included in Ansible. (Also it will reduce package bloat.)

+1, that's true, as we saw, the main concern for many maintainers was "where will the collection appear on docs.ansible.com"

>One big question though is a) how this should work, and b) who does all the certifications.

My vision is that it'll be the same as the inclusion process. As i mentioned above, we could just replace `inclusion` with `community certification` everywhere.

I'm not against inclusion as-is. The suggested "certification" model imo brings the benefits like not shipping the kitchensink keeping recognition, visibility and motivation as they are not.

I have a strong feeling that we should have only one model: inclusion or certification, not both co-existing.

>TLDR: I like the idea, but it needs more work. But IMO it's a good idea to start discussing it now and keeping it in the backs of our minds, so we can remove these obstacles eventually.

+1

**Comment by briantist** [comment]
> I think this is the new one:)

Thanks @Andersson007 , before I respond to those points, are we sure this is the place we want to discuss that? 
I really think that this deserves to be its own topic, or that we should re-open the existing one, if we want to bring this up again.

**Comment by Andersson007** [comment]
> > I think this is the new one:)
> 
> Thanks @Andersson007 , before I respond to those points, are we sure this is the place we want to discuss that? I really think that this deserves to be its own topic, or that we should re-open the existing one, if we want to bring this up again.

@briantist i believe we should agree/disagree on the alternative first (i.e. on the collection community certification instead of inclusion) before re-opening / creating another topic. I.e. we should show the community a concrete thing so that they'll see well how it can work. Would be imo more effective. What do you think?

**Comment by briantist** [comment]
> i believe we should agree/disagree on the alternative first (i.e. on the collection community certification instead of inclusion) before re-opening / creating another topic. I.e. we should show the community a concrete thing so that they'll see well how it can work. Would be imo more effective. What do you think?

One of the reasons I want to have the other discussion first is that it's difficult to evaluate whether an alternative is viable without understanding why the current state of things is not sufficient. To me it seems better to have that discussion in isolation, because there are so many different feelings and opinions on that. 

Without a clear consensus (not necessarily full agreement) on the future of the package and the issues we are trying to address, all discussions about alternatives will turn into that discussion :)

---

But let's let some other people weigh in, I am fine to discuss that here if that's what is preferred!

**Comment by mariolenz** [comment]
> One big question though is a) how this should work, and b) who does all the certifications. Regarding b), collection reviews take a lot of time. We are already struggling to get enough reviews for collections that want to be included in Ansible, so opening another can of worms^Hreview request sources sounds problematic.

Very good point. Personally, I think we won't be able to put a lot of manual work in certifying collections. But what if we could automate this?

I mean, it all depends on the requirements for the certification. Let's say: A collection can get certified if it succeeds the sanity tests of the current ansible-core release. This is something that can be checked automatically. We could set up a CI pipeline to test this. We could even run this regularly and revoke certification automatically if this starts to fail for a collection.

I'm not saying that succeeding sanity tests should be the only requirement, this is just an example. But I think we should try to automate this. Otherwise, we will strugle to get reviews as we are already struggling to get reviews for collections that want to be included in Ansible.

So the certification could be something between "nothing at all" and "inclusion in the Ansible package". As an idea, we could also require collections that want to be included into the Ansible package to be certified. If we can automate the checks for certification, the inclusion requirements might be simplified. We would only have to check things that are hard to or can't be automated.

**Comment by samccann** [comment]
A few comments:

1. Not sure we should start community certification. That could easily be confused with Red Hat certifications, yet the two would not be using the same set of requirements, for example.
2. galaxy content is findable on google. Search `ansible docker role` and the galaxy result shows up on the first page. Not ideal but  that's likely because it's popular and other web hits get more attention than the galaxy one does. Or search ansible virtualization collection and up pops the collection on galaxy. 
3. Docs on galaxy will be available on galaxyng which is coming soon. So not only will a user find the collection, they will be able to read the collection-level documentation.

This doesn't solve the 'offline' folks who need to get ansible etc through a mirror (and pypi is the easiest to get on those mirrors). There seems to be a couple of options that come to mind:
A. Keep letting the Ansible package growing.  It seems all we are suggesting that's different now is the potential to [allow inclusion of role-based collections](https://github.com/ansible-community/community-topics/issues/197). If we see the floodgates opening, we may need to change strategy, but until we see more new collections per release than we can handle, maybe this isn't a problem (yet).
B. We create a new ansible-extras pypi package and stop expanding the existing Ansible package. We use the same criteria we have today, but it goes to a new bundle so to speak instead of the existing bundle. Some smart person could figure out how to spin up a docsite for this package that is potentially easier to maintain than docs.ansible.com.

**Comment by Andersson007** [comment]
@samccann

> Not sure we should start community certification. That could easily be confused with Red Hat certifications, yet the two would not be using the same set of requirements, for example.

`Community certification` is just a term that came to mind first, maybe not the best one. We could name it `Community approved collections` or whatever. I'll update the description.

> If we see the floodgates opening, we may need to change strategy, but until we see more new collections per release than we can handle, maybe this isn't a problem (yet).

Ability to handle future role-focused collection requests is a good question too.
The primary goal of this topic imo is to decide whether `Approved` status is sufficient, not tarball physical presence.
Even if the collection docs will be available on Galaxy soon, there can be list of community approved collections on docs.ansible.com which would distinguish them from all other collections.

Speaking about folks with local isolated environments like in case of @mariolenz , if they need a role or a collection that is not included in the PyPI package, what will they do? Will it be a blocker so that they have to write all not included roles/collections ourselves, i.e. to reinvent the wheel?
Is it really hard to assemble your own custom package or just download the set of collections needed and copy them to controllers?

**Comment by Andersson007** [comment]
Seems that the Community is not interested in switching to this scheme right now.
Let's keep it in mind though if the situation changes.
I'm closing this now to keep the tracker clean. Feel free to comment/re-open if you want to discuss this.
Thanks everyone
